src/plugins/dashboard/public/application/hooks/use_dashboard_app_state.ts

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import _ from 'lodash';
+import { History } from 'history';
+import { debounceTime } from 'rxjs/operators';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { BehaviorSubject, combineLatest, Subject } from 'rxjs';
+
+import { DashboardConstants } from '../..';
+import { ViewMode } from '../../services/embeddable';
+import { useKibana } from '../../services/kibana_react';
+import { getNewDashboardTitle } from '../../dashboard_strings';
+import { IKbnUrlStateStorage } from '../../services/kibana_utils';
+import { setDashboardState, useDashboardDispatch, useDashboardSelector } from '../state';
+import {
+  DashboardBuildContext,
+  DashboardAppServices,
+  DashboardAppState,
+  DashboardRedirect,
+  DashboardState,
+} from '../../types';
+import {
+  tryDestroyDashboardContainer,
+  syncDashboardContainerInput,
+  savedObjectToDashboardState,
+  syncDashboardIndexPatterns,
+  syncDashboardFilterState,
+  loadSavedDashboardState,
+  buildDashboardContainer,
+  loadDashboardUrlState,
+  diffDashboardState,
+  areTimeRangesEqual,
+} from '../lib';
+
+export interface UseDashboardStateProps {
+  history: History;
+  savedDashboardId?: string;
+  redirectTo: DashboardRedirect;
+  isEmbeddedExternally: boolean;
+  kbnUrlStateStorage: IKbnUrlStateStorage;
+}
+
+export const useDashboardAppState = ({
+  history,
+  redirectTo,
+  savedDashboardId,
+  kbnUrlStateStorage,
+  isEmbeddedExternally,
+}: UseDashboardStateProps) => {
+  const dispatchDashboardStateChange = useDashboardDispatch();
+  const dashboardState = useDashboardSelector((state) => state.dashboardStateReducer);
+
+  /**
+   *  Dashboard app state is the return value for this hook and contains interaction points that the rest of the app can use
+   * to read or manipulate dashboard state.
+   */
+  const [dashboardAppState, setDashboardAppState] = useState<DashboardAppState>(() => ({
+    $onDashboardStateChange: new BehaviorSubject({} as DashboardState),
+    $triggerDashboardRefresh: new Subject<{ force?: boolean }>(),
+  }));
+
+  /**
+   *  Last saved state is diffed against the current dashboard state any time either changes. This is used to set the
+   * unsaved changes portion of the dashboardAppState.
+   */
+  const [lastSavedState, setLastSavedState] = useState<DashboardState>();
+  const $onLastSavedStateChange = useMemo(() => new Subject<DashboardState>(), []);
+
+  /**
+   * Unpack services
+   */
+  const services = useKibana<DashboardAppServices>().services;
```

Comment:
Here, I've moved the unpacking of the services outside the useEffect. This has the downside of adding far more individual items to the deps array, but stops the useEffect from monitoring unnecessary services.